documentation/architecture.md

```diff
@@ -0,0 +1,337 @@
+same time, it needs to be able to associate any given stream to a container process. This is done by `hyperstart`
+allocating a set of at most 2 so-called sequence numbers per container process. `cc-oci-runtime` will send
+the `AllocateIO` command to `cc-proxy` to have it request `hyperstart` to allocate those sequence numbers.
+They will be passed as command line arguments to `cc-shim`, who will then use them to e.g. prepend its stdin
+stream packets with the right sequence number.
+- `Hyper`: This command is used by both `cc-oci-runtime` and `cc-shim` to forward `hyperstart` specific
+commands.
+
+For more details about `cc-proxy`'s protocol, theory of operations or debugging tips, please read
+[`cc-proxy` README](https://github.com/01org/cc-oci-runtime/tree/master/proxy).
+
+#### Shim
+
+Docker's `containerd-shim` is designed around the assumption that it can monitor and reap the actual
+container process. As `containerd-shim` runs on the host, it can not directly monitor a process running
+within a virtual machine. At most it can see the QEMU process, but that is not enough.
+With Clear Containers, `cc-shim` acts as the container process that `containerd-shim` can monitor. Therefore
+`cc-shim` needs to handle all container I/O streams (`stdout`, `stdin` and `stderr`) and forward all signals
+`containerd-shim` decides to send to the container process.
+
+`cc-shim` has an implicit knowledge about which VM agent will handle those streams and signals and thus act as
+an encapsulation layer between `containerd-shim` and `hyperstart`:
+
+- It fragments and encapsulates the standard input stream from containerd-shim into `hyperstart` stream packets:
+```
+  ┌─────────────────┬───────────────┬────────────────────┐
+  │    IO stream    │ Packet length │ IO stream fragment │
+  │ sequence number │    (32 bits)  │                    │
+  └─────────────────┴───────────────┴────────────────────┘
+```
+- It de-encapsulates and assembles standard output and error `hyperstart` stream packets into an output stream
+that it forwards to `containerd-shim`
+- It translates all UNIX signals (except `SIGKILL` and `SIGSTOP`) into `hyperstart` `KILLCONTAINER` commands
+that it sends to the VM via `cc-proxy` UNIX named socket.
+
+The IO stream sequence numbers are passed from `cc-proxy` to `cc-shim` when the former spawns the latter.
+They are generated by `hyperstart` and `cc-oci-runtime` fetches them by sending the `AllocateIO` command to
+`cc-proxy`.
+
+As an example, let's say that running the `# pwd` command from a container standard input will generate
+`/tmp` from the container standard output. `hyperstart` assigned this specific process 8888 and 8889 respectively
+as the stdin, stdour and stderr sequence numbers.
+With `cc-shim` and Clear Containers, this example would look like:
+
+![cc-shim](https://github.com/01org/cc-oci-runtime/blob/topic/documentation/documentation/shim.png)
```

Comment:
I appreciate this is mentioned in the text, but -- although it might be difficult to squeeze this in -- I think it would help if you could add a visual hint that those 3 boxes are hyperstart packets. Maybe a blue border along with an entry in the key below stdout and stdin would make that clearer (or even just adding "seq", "len" and "data" above each part of the packets).